Jersey 2.25 crashes with NPE on startup

> java.lang.RuntimeException: org.kantega.reststop.core.InvalidPluginException: Plugin class org.kantega.reststop.jaxrs.JerseyPlugin threw an exeception during construction 
> 	at org.eclipse.jetty.annotations.ServletContainerInitializersStarter.doStart(ServletContainerInitializersStarter.java:68)
> 	at org.eclipse.jetty.util.component.AbstractLifeCycle.start(AbstractLifeCycle.java:68)
> 	at org.eclipse.jetty.servlet.ServletContextHandler.startContext(ServletContextHandler.java:330)
> 	at org.eclipse.jetty.webapp.WebAppContext.startWebapp(WebAppContext.java:1404)
> 	at org.eclipse.jetty.maven.plugin.JettyWebAppContext.startWebapp(JettyWebAppContext.java:323)
> 	at org.eclipse.jetty.webapp.WebAppContext.startContext(WebAppContext.java:1366)
> 	at org.eclipse.jetty.server.handler.ContextHandler.doStart(ContextHandler.java:772)
> 	at org.eclipse.jetty.servlet.ServletContextHandler.doStart(ServletContextHandler.java:262)
> 	at org.eclipse.jetty.webapp.WebAppContext.doStart(WebAppContext.java:520)
> 	at org.eclipse.jetty.maven.plugin.JettyWebAppContext.doStart(JettyWebAppContext.java:398)
> 	at org.eclipse.jetty.util.component.AbstractLifeCycle.start(AbstractLifeCycle.java:68)
> 	at org.kantega.reststop.maven.AbstractReststopRunMojo.startJetty(AbstractReststopRunMojo.java:139)
> 	at org.kantega.reststop.maven.AbstractReststopRunMojo.execute(AbstractReststopRunMojo.java:81)
> 	at org.apache.maven.plugin.DefaultBuildPluginManager.executeMojo(DefaultBuildPluginManager.java:134)
> 	at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:207)
> 	at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:153)
> 	at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:145)
> 	at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject(LifecycleModuleBuilder.java:116)
> 	at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject(LifecycleModuleBuilder.java:80)
> 	at org.apache.maven.lifecycle.internal.builder.singlethreaded.SingleThreadedBuilder.build(SingleThreadedBuilder.java:51)
> 	at org.apache.maven.lifecycle.internal.LifecycleStarter.execute(LifecycleStarter.java:128)
> 	at org.apache.maven.DefaultMaven.doExecute(DefaultMaven.java:307)
> 	at org.apache.maven.DefaultMaven.doExecute(DefaultMaven.java:193)
> 	at org.apache.maven.DefaultMaven.execute(DefaultMaven.java:106)
> 	at org.apache.maven.cli.MavenCli.execute(MavenCli.java:863)
> 	at org.apache.maven.cli.MavenCli.doMain(MavenCli.java:288)
> 	at org.apache.maven.cli.MavenCli.main(MavenCli.java:199)
> 	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
> 	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
> 	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
> 	at java.lang.reflect.Method.invoke(Method.java:498)
> 	at org.codehaus.plexus.classworlds.launcher.Launcher.launchEnhanced(Launcher.java:289)
> 	at org.codehaus.plexus.classworlds.launcher.Launcher.launch(Launcher.java:229)
> 	at org.codehaus.plexus.classworlds.launcher.Launcher.mainWithExitCode(Launcher.java:415)
> 	at org.codehaus.plexus.classworlds.launcher.Launcher.main(Launcher.java:356)
> Caused by: org.kantega.reststop.core.InvalidPluginException: Plugin class org.kantega.reststop.jaxrs.JerseyPlugin threw an exeception during construction 
> 	at org.kantega.reststop.core.ReststopPluginLoader.lambda$loadPlugin$0(ReststopPluginLoader.java:70)
> 	at org.kantega.reststop.core.ReststopPluginLoader.withClassloaderContext(ReststopPluginLoader.java:266)
> 	at org.kantega.reststop.core.ReststopPluginLoader.loadPlugin(ReststopPluginLoader.java:61)
> 	at org.kantega.reststop.core.PluginDeployer.deploy(PluginDeployer.java:46)
> 	at org.kantega.reststop.core.PluginDeployer.deploy(PluginDeployer.java:104)
> 	at org.kantega.reststop.core.DefaultReststopPluginManager.deploy(DefaultReststopPluginManager.java:52)
> 	at org.kantega.reststop.servlets.ReststopInitializer.deployPlugins(ReststopInitializer.java:89)
> 	at org.kantega.reststop.servlets.ReststopInitializer.onStartup(ReststopInitializer.java:79)
> 	at org.eclipse.jetty.plus.annotation.ContainerInitializer.callStartup(ContainerInitializer.java:140)
> 	at org.eclipse.jetty.annotations.ServletContainerInitializersStarter.doStart(ServletContainerInitializersStarter.java:63)
> 	... 34 more
> Caused by: java.lang.reflect.InvocationTargetException
> 	at sun.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
> 	at sun.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:62)
> 	at sun.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
> 	at java.lang.reflect.Constructor.newInstance(Constructor.java:423)
> 	at org.kantega.reststop.core.ReststopPluginLoader.lambda$loadPlugin$0(ReststopPluginLoader.java:63)
> 	... 43 more
> Caused by: java.lang.NullPointerException
> 	at org.glassfish.jersey.servlet.init.FilterUrlMappingsProviderImpl.getFilterUrlMappings(FilterUrlMappingsProviderImpl.java:66)
> 	at org.glassfish.jersey.servlet.ServletContainer.init(ServletContainer.java:444)
> 	at org.kantega.reststop.jaxrs.JerseyPlugin.<init>(JerseyPlugin.java:59)
> 	... 48 more
>

--- a/plugins/jersey/src/main/java/org/kantega/reststop/jaxrs/JerseyPlugin.java
+++ b/plugins/jersey/src/main/java/org/kantega/reststop/jaxrs/JerseyPlugin.java
@@ -20,21 +20,24 @@ import org.glassfish.jersey.jackson.JacksonFeature;
 import org.glassfish.jersey.server.ResourceConfig;
 import org.glassfish.jersey.servlet.ServletContainer;
 import org.glassfish.jersey.servlet.ServletProperties;
-import org.kantega.reststop.api.*;
+import org.kantega.reststop.api.Export;
+import org.kantega.reststop.api.Plugin;
 import org.kantega.reststop.jaxrsapi.ApplicationBuilder;
 import org.kantega.reststop.jaxrsapi.ApplicationDeployer;
 import org.kantega.reststop.servlet.api.FilterPhase;
 import org.kantega.reststop.servlet.api.ServletBuilder;
 
-import javax.servlet.Filter;
-import javax.servlet.FilterChain;
-import javax.servlet.ServletException;
+import javax.servlet.*;
 import javax.servlet.http.HttpServletRequest;
 import javax.servlet.http.HttpServletResponse;
 import javax.ws.rs.core.Application;
 import javax.ws.rs.core.MediaType;
 import java.io.IOException;
+import java.lang.reflect.Proxy;
 import java.util.*;
+
+import static java.util.Collections.emptyList;
+import static java.util.Collections.singletonList;
 
 /**
  *
@@ -54,12 +57,22 @@ public class JerseyPlugin implements ApplicationDeployer, ApplicationBuilder {
 
     public JerseyPlugin(final ServletBuilder reststop) throws ServletException {
 
-        filter = addJerseyFilter(new ReststopApplication(Collections.emptyList()));
+        filter = addJerseyFilter(new ReststopApplication(emptyList()));
 
-        filter.init(reststop.filterConfig("jersey", new Properties()));
+        String pathMapping = "/*";
+        filter.init(createFilterConfig(reststop, pathMapping));
 
-        jerseyFilter = reststop.filter(filter, FilterPhase.USER, "/*");
+        jerseyFilter = reststop.filter(filter, FilterPhase.USER, pathMapping);
     }
+
+    private FilterConfig createFilterConfig(ServletBuilder reststop, String pathMapping) {
+        String filterName = "jersey";
+        FilterConfig filterConfig = reststop.filterConfig(filterName, new Properties());
+
+        return new MappingAwareFilterConfig(filterConfig, filterName, pathMapping);
+    }
+
+
 
     private ServletContainer addJerseyFilter(Application application) {
         ResourceConfig resourceConfig = getResourceConfig(application);
@@ -135,5 +148,98 @@ public class JerseyPlugin implements ApplicationDeployer, ApplicationBuilder {
                 };
             }
         };
+    }
+
+    private static class MappingAwareFilterConfig implements FilterConfig {
+        private final FilterConfig filterConfig;
+        private final ServletContext servletContext;
+
+        MappingAwareFilterConfig(FilterConfig filterConfig, String filterName, String pathMapping) {
+            this.filterConfig = filterConfig;
+            this.servletContext = createMappingAwareServletContext(filterConfig.getServletContext(), filterName, pathMapping);
+        }
+
+        @Override
+        public String getFilterName() {
+            return filterConfig.getFilterName();
+        }
+
+        @Override
+        public ServletContext getServletContext() {
+            return servletContext;
+        }
+
+        @Override
+        public String getInitParameter(String name) {
+            return filterConfig.getInitParameter(name);
+        }
+
+        @Override
+        public Enumeration<String> getInitParameterNames() {
+            return filterConfig.getInitParameterNames();
+        }
+
+        private ServletContext createMappingAwareServletContext(ServletContext filterConfig, String filterName, String pathMapping) {
+            return (ServletContext) Proxy.newProxyInstance(getClass().getClassLoader(), new Class[]{ServletContext.class}, (proxy1, method, args) -> {
+                if(method.getName().equals("getFilterRegistration") && args.length == 1 && filterName.equals(args[0])) {
+                    return createFilterRegistration(pathMapping);
+                }
+                return method.invoke(filterConfig, args);
+            });
+        }
+
+        private FilterRegistration createFilterRegistration(final String pathMapping) {
+            return new FilterRegistration() {
+                @Override
+                public void addMappingForServletNames(EnumSet<DispatcherType> dispatcherTypes, boolean isMatchAfter, String... servletNames) {
+                    throw new IllegalStateException("This method should not be called");
+                }
+
+                @Override
+                public Collection<String> getServletNameMappings() {
+                    throw new IllegalStateException("This method should not be called");
+                }
+
+                @Override
+                public void addMappingForUrlPatterns(EnumSet<DispatcherType> dispatcherTypes, boolean isMatchAfter, String... urlPatterns) {
+                    throw new IllegalStateException("This method should not be called");
+                }
+
+                @Override
+                public Collection<String> getUrlPatternMappings() {
+                    return singletonList(pathMapping);
+                }
+
+                @Override
+                public String getName() {
+                    throw new IllegalStateException("This method should not be called");
+                }
+
+                @Override
+                public String getClassName() {
+                    throw new IllegalStateException("This method should not be called");
+                }
+
+                @Override
+                public boolean setInitParameter(String name, String value) {
+                    throw new IllegalStateException("This method should not be called");
+                }
+
+                @Override
+                public String getInitParameter(String name) {
+                    throw new IllegalStateException("This method should not be called");
+                }
+
+                @Override
+                public Set<String> setInitParameters(Map<String, String> initParameters) {
+                    throw new IllegalStateException("This method should not be called");
+                }
+
+                @Override
+                public Map<String, String> getInitParameters() {
+                    throw new IllegalStateException("This method should not be called");
+                }
+            };
+        }
     }
 }

--- a/plugins/security/src/main/java/org/kantega/reststop/security/BasicAuthFilter.java
+++ b/plugins/security/src/main/java/org/kantega/reststop/security/BasicAuthFilter.java
@@ -22,6 +22,7 @@ import javax.servlet.http.HttpServletRequestWrapper;
 import javax.servlet.http.HttpServletResponse;
 import javax.xml.bind.DatatypeConverter;
 import java.io.IOException;
+import java.security.Principal;
 
 /**
  *
@@ -47,6 +48,11 @@ public class BasicAuthFilter implements Filter {
                     @Override
                     public String getRemoteUser() {
                         return usernameAndPassword[0];
+                    }
+
+                    @Override
+                    public Principal getUserPrincipal() {
+                        return () -> usernameAndPassword[0];
                     }
 
                     @Override

--- a/pom.xml
+++ b/pom.xml
@@ -55,7 +55,7 @@
 
     <properties>
         <project.build.sourceEncoding>UTF-8</project.build.sourceEncoding>
-        <jersey.version>2.17</jersey.version>
+        <jersey.version>2.25</jersey.version>
         <jetty.version>9.3.9.v20160517</jetty.version><!-- Also defined in template-plugin-webapp-pom.xml -->
         <maven.version>3.1.0</maven.version>
         <maven-deploy-plugin.version>2.8.1</maven-deploy-plugin.version>

